### DISCUSSION TITLE
How to detect objects in real time with a cv2 image?

### DISCUSSION THREAD

**Discussion by shershunov** [ultralytics/yolov5#2803] [Q&A] [answered]
I want to detect objects in a cv2 image in real time. I have a code that captures the screen using the mfc library. How do I send an image to yolov5 at maximum speed? If you can help me, I will be very grateful!

```py
from cv2 import imshow, waitKey, destroyAllWindows
from mss import mss
from numpy import array

with mss() as sct:
    monitor = {"top": 0, "left": 0, "width": 600, "height": 600}
    while True:
        img = array(sct.grab(monitor))
        imshow("Detect", img)
        if waitKey(1) & 0xFF == ord("q"):
            destroyAllWindows()
            break
```

**Comment by glenn-jocher** [comment, marked as answer]
@HamkiWhite you can use YOLOv5 PyTorch Hub models for inference in your loop.

### PyTorch Hub

Inference with YOLOv5 and [PyTorch Hub](https://github.com/ultralytics/yolov5/issues/36):
```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')

# Image
img = 'https://github.com/ultralytics/yolov5/raw/master/data/images/zidane.jpg'

# Inference
results = model(img)
```

**Reply by shershunov** [reply]
How do I view an image in real time using cv2.imshow()?

**Reply by 9christian9** [reply]
I've seen you screenshot a few minutes ago and the proble is that you use directly the model​(​img​) that is wrong...
I'm new of this world and I learning it so sorry if my reply is not very detailed.

Try this:
```py
# Im is taken from video
model = model(im, size=640)
xywh = imodel.xyxy[0]
xywh = torch.tensor(xywh)[0]
x1 = int(xywh[0])
x2 = int(xywh[1])
x3 = int(xywh[2])
x4 = int(xywh[3])
cv2.rectangle(im, (x1, x2), (x3, x4), (0, 255, 0), 3)
cv2.imshow('show', im)
```

**Reply by shershunov** [reply]
Thank you, it works! But if the object is not visible, the application crashes...

**Reply by 9christian9** [reply]
It is normal because my script not manege the value null inside Tensor....
You can try with the yolo.pt and using your webcam... if yolo recognise you, the script not end. When yolo not recognise you, the script end with an error...

**Reply by shershunov** [reply]
I fixed the error, but how do I make it so that multiple objects are displayed? Only one is currently displayed(

**Reply by 9christian9** [reply]
Can you put your fix code here?
So it is normal the found only an object because I've used  a static point xywh[0]...
I think that you should be do a cycle of Tensor from position 0 to position 3 that are X1 Y1 X2 Y2... for any Tensor detected...
In this moment I write you from my telephone and I've not tested it...

**Reply by shershunov** [reply]
```py
from cv2 import waitKey, destroyAllWindows, imshow, rectangle
from mss import mss
from numpy import array
from win32gui import FindWindow, GetWindowRect
import torch

with mss() as sct:
    # hwnd = {"top": 40, "left": 0, "width": 800, "height": 640}
    hwnd = FindWindow(None, r'Counter-Strike: Global Offensive')
    model = torch.hub.load('ultralytics/yolov5', 'custom', path_or_model='best.pt')
    model = model.autoshape()
    while True:
        dimensions = GetWindowRect(hwnd)
        img = array(sct.grab(dimensions))
        result = model(img, size=640)
        pos = result.xyxy[0]
        if len(pos) > 0:
            pos = torch.tensor(pos)[0]
            x = int(pos[0])
            y = int(pos[1])
            w = int(pos[2])
            h = int(pos[3])
            rectangle(img, (x, y), (w, h), (255, 255, 255), 1)
        imshow('Detect', img)
        if waitKey(1) & 0xFF == ord("q"):
            destroyAllWindows()
            break
```
![изображение](https://user-images.githubusercontent.com/71601841/115070946-47e4f080-9efe-11eb-9ddb-f61f816d6ba0.png)

**Reply by 9christian9** [reply]
Try to print model(img) please...
If you print it can you see al Tensor detected in your picture...

**Reply by shershunov** [reply]
He sees all the opponents, but the frame is drawn only near one.

**Reply by 9christian9** [reply]
You should be do:
- For each Tensor
-  draw a rectangle
- after have drew all rectangles put in the image output

This is the theory 😆

**Reply by shershunov** [reply]
Thank)

**Reply by 9christian9** [reply]
if you have success, share the code that always is good

**Reply by shershunov** [reply]
Ok)

**Reply by shershunov** [reply]
```py
img = array(sct.grab(dimensions))
        result = model(img, size=640)
        repos = result.xyxy[0]
        if len(repos) > 0:
            a = 0
            for i in range(len(repos)):
                pos = torch.tensor(repos)[a]
                rectangle(img, (int(pos[0]), int(pos[1])), (int(pos[2]), int(pos[3])), (255, 255, 255), 2)
                a += 1
        imshow('Detect', img)
```

**Reply by 9christian9** [reply]
@EkelviNistars 
I've seen that .detach().clone() is the best practice for clone a tensor
I've removed the variable a to control a cycle... try to use i inside the cycle
I suggest changing your code in:

```python
    result = model(img, size=640)
    repos = result.xyxy[0]
    if len(repos) > 0:
        for i in range(len(repos)):
            pos = repos[i].detach().clone()
            rectangle(img, (int(pos[0]), int(pos[1])), (int(pos[2]), int(pos[3])), (255, 255, 255), 2)
    imshow('Detect', img)
```
